community/bolt/src/main/java/org/neo4j/bolt/runtime/DefaultBoltConnection.java

```diff
@@ -0,0 +1,314 @@
+                                }
+                                else
+                                {
+                                    machine.validateTransaction();
+                                }
+                            }
+                        }
+                        notifyDrained( batch );
+
+                        while ( batch.size() > 0 )
+                        {
+                            Job current = batch.remove( 0 );
+
+                            current.perform( machine );
+                        }
+
+                        loop = machine.shouldStickOnThread();
+                        waitForMessage = loop;
+                    }
+                }
+            }
+            while ( loop );
+
+            assert !machine.hasOpenStatement();
+        }
+        catch ( BoltConnectionAuthFatality ex )
+        {
+            // do not log
+            shouldClose.set( true );
+        }
+        catch ( BoltProtocolBreachFatality ex )
+        {
+            shouldClose.set( true );
+            log.error( String.format( "Protocol breach detected in bolt session '%s'.", id() ), ex );
+        }
+        catch ( InterruptedException ex )
+        {
+            shouldClose.set( true );
+            log.info( "Bolt session '%s' is interrupted probably due to server shutdown.", id() );
+        }
+        catch ( Throwable t )
+        {
+            shouldClose.set( true );
+            userLog.error( String.format( "Unexpected error detected in bolt session '%s'.", id() ), t );
+        }
+        finally
+        {
+            if ( shouldClose.get() )
+            {
+                close();
+            }
+        }
+
+        return !closed.get();
+    }
+
+    @Override
+    public void handleSchedulingError( Throwable t )
+    {
+        String message;
+        Neo4jError error;
+        if ( ExceptionUtils.hasCause( t, RejectedExecutionException.class ) )
+        {
+            error = Neo4jError.from( Status.Request.NoThreadsAvailable, Status.Request.NoThreadsAvailable.code().description() );
+            message = String.format( "Unable to schedule bolt session '%s' for execution since there are no available threads to " +
+                    "serve it at the moment. You can retry at a later time or consider increasing max pool / queue size for bolt connector(s).", id() );
+        }
+        else
+        {
+            error = Neo4jError.fatalFrom( t );
+            message = String.format( "Unexpected error during scheduling of bolt session '%s'.", id() );
+        }
+
+        log.error( message, t );
+        userLog.error( message );
+        machine.markFailed( error );
+        processNextBatch( 1 );
+    }
+
+    @Override
+    public void interrupt()
+    {
+        machine.interrupt();
+    }
+
+    @Override
+    public void stop()
+    {
+        if ( shouldClose.compareAndSet( false, true ) )
+        {
+            machine.terminate();
+
+            if ( !hasPendingJobs() )
```

Comment:
I think it would be even better not to break the linearity of operations on a bolt connection, and changed this to enqueue an empty job to the queue and have it executed by the thread pool. WDYT?